ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -832,12 +875,47 @@ private JoinKey buildForeignJoinKey(final Join join, final ColumnReferenceExp co
 
     final ColumnName foreignKeyColumnName = columnRef.maybeQualifier().isPresent()
         ? ColumnNames.generatedJoinColumnAlias(
-            columnRef.maybeQualifier().get(), columnRef.getColumnName())
+            columnRef.maybeQualifier().get(),
+            columnRef.getColumnName())
         : columnRef.getColumnName();
 
     return JoinKey.foreignKeyColumn(foreignKeyColumnName, leftSourceKeys);
   }
 
+  private static void verifyJoinConditionTypes(
+      final SqlType leftType,
+      final SqlType rightType,
+      final Expression leftExpression,
+      final Expression rightExpression,
+      final boolean flipped
+  ) {
+    if (!leftType.equals(rightType)) {
+      final SqlType maybeFlippedLeftType;
+      final SqlType maybeFlippedRightType;
+      final Expression maybeFlippedLeftExpression;
+      final Expression maybeFlippedRightExpression;
+      if (flipped) {
+        maybeFlippedLeftType = rightType;
+        maybeFlippedRightType = leftType;
+        maybeFlippedLeftExpression = rightExpression;
+        maybeFlippedRightExpression = leftExpression;
+      } else {
+        maybeFlippedLeftType = leftType;
+        maybeFlippedRightType = rightType;
+        maybeFlippedLeftExpression = leftExpression;
+        maybeFlippedRightExpression = rightExpression;
+      }
+
+      throw new KsqlException(String.format(
```

Comment:
nit: just use a ternary operator inside `String.format()` here?

Reply:
Not sure what you mean?

Reply:
Oh I just meant maybe we do not need to declare these local fields, and just inside `String.format` we do:

```
...
flipped ? rightExpression: leftExpression,
flipped ? rightType: leftType,
flipped ? leftExpression: rightExpression,
flipped ? leftType: rightType,
```